src/operator/contrib/sync_batch_norm-inl.h

```diff
@@ -0,0 +1,592 @@
+template<class T>
+class SharedND {
+ private:
+  int num_devices_;
+  T mean_;
+  T *data_;
+  bool *flag_;
+  bool mean_ready_ = false;
+  bool data_inited_ = false;
+  std::mutex mutex_;
+
+ public:
+  explicit SharedND(int ndev) :num_devices_(ndev) {
+    flag_ = new bool[ndev];
+    data_ = new T[ndev];
+    memset(flag_, false, ndev * sizeof(bool));
+  }
+
+  ~SharedND() {
+    mshadow::FreeSpace(&mean_);
+    delete [] flag_;
+    delete [] data_;
+  }
+
+  void Init(mshadow::Shape<1> shape) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (!data_inited_) {
+      for (int i = 0; i < num_devices_; i++) {
+        data_[i] = mshadow::NewTensor<cpu, real_t>(shape, 0.0f);
+      }
+      mean_ = mshadow::NewTensor<cpu, real_t>(shape, 0.0f);
+      data_inited_ = true;
+    }
+  }
+
+  T* Retrieve(mshadow::Shape<1> shape, int index) {
+    if (!data_inited_) {
+      Init(shape);
+    }
+    if (flag_[index] == false) {
+      return &data_[index];
+    } else {
+      return nullptr;
+    }
+  }
+
+  bool SetReady(int index) {
+    if (flag_[index] == false) {
+      flag_[index] = true;
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  T Pop(int index) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    while (!MeanReady()) {}
+    flag_[index] = false;
+    T tmp = mean_;
+    ResetMean();
+    return tmp;
+  }
+
+  bool MeanReady() {
+    if (mean_ready_) {
+      return true;
+    }
+    for (int i = 0; i < num_devices_; i++) {
+      if (!flag_[i]) {
+        return false;
+      }
+    }
+    for (int i = 1; i < num_devices_; i++) {
+      data_[0] += data_[i];
+    }
+    mean_ = data_[0] * 1.0f /  num_devices_;
+    mean_ready_ = true;
+    return true;
+  }
+
+  void ResetMean() {
+    for (int i = 0; i < num_devices_; i++) {
+      if (flag_[i]) return;
+    }
+    mean_ready_ = false;
+  }
+};
+
+template<class T>
+class GlobalShared {
+ public:
+  T* Register(const std::string &key, int ndev) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    auto it = registry_.find(key);
+    if (it != registry_.end()) return it->second;
+    T *newT = new T(ndev);
+    registry_[key] = newT;
+    return newT;
+  }
+  ~GlobalShared() {
+    for (auto it = registry_.begin(); it != registry_.end(); it++) {
+      T *ptr = it->second;
+      delete ptr;
+    }
+  }
+ private:
+  std::mutex mutex_;
+  std::map<std::string, T*> registry_;
+};
+
+template<class T>
+class GlobalSharedRank {
+ public:
+  T Register(const std::string &key, int ndev) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    auto it = registry_.find(key);
+    if (it != registry_.end()) {
+      T* tmpT = it->second;
+      *tmpT = (*tmpT == ndev - 1) ? 0 : *tmpT + 1;
+      return *tmpT;
+    }
+    T *newT = new T(0);
+    registry_[key] = newT;
+    return *newT;
+  }
+  ~GlobalSharedRank() {
+    for (auto it = registry_.begin(); it != registry_.end(); it++) {
+      T *ptr = it->second;
```

Comment:
same here

Reply:
If not inited, the hash map should be empty

Reply:
I see. One more thing, you should lock guard the destructor

Reply:
ok, should be fine